lib/api/v3/principals/principal_representer_factory.rb

```diff
@@ -0,0 +1,54 @@
+#-- encoding: UTF-8
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2021 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+module API
+  module V3
+    module Principals
+      class PrincipalRepresenterFactory
+
+        ##
+        # Create the appropriate subclass representer
+        # for each principal entity
+        def self.create(model, *args)
+          case model.type
+          when 'User'
+            ::API::V3::Users::UserRepresenter.new(model, *args)
+          when 'Group'
+            ::API::V3::Groups::GroupRepresenter.new(model, *args)
+          when 'PlaceholderUser'
+            ::API::V3::PlaceholderUsers::PlaceholderUserRepresenter.new(model, *args)
```

Comment:
For the time being, I'd keep it at calling the `#create` method. While the representers currently do not include custom fields, they should and when that is finally implemented, having to find this spot again will be hard to do.

:+: for creating this file, it can now also be used in https://github.com/opf/openproject/blob/dev/lib/api/v3/principals/associated_subclass_lambda.rb#L81